src/clients/java/src/main/java/com/tigerbeetle/NativeClient.java

```diff
@@ -31,31 +36,51 @@ private static void assertArgs(final byte[] clusterID, final String addresses,
-        if (contextHandle == 0L)
-            throw new IllegalStateException("Client is closed");
-
-        final var packet_acquire_status = submit(contextHandle, request);
-        if (packet_acquire_status == PacketAcquireStatus.ConcurrencyMaxExceeded.value) {
-            throw new ConcurrencyExceededException();
-        } else if (packet_acquire_status == PacketAcquireStatus.Shutdown.value) {
-            throw new IllegalStateException("Client is closing");
-        } else {
-            assertTrue(packet_acquire_status == PacketAcquireStatus.Ok.value,
-                    "PacketAcquireStatus=%d is not implemented", packet_acquire_status);
+        try {
+            contextHandleReferences.incrementAndGet();
+            final var handle = contextHandle.getAcquire();
+
+            if (handle == 0L)
+                throw new IllegalStateException("Client is closed");
+
+            final var packet_acquire_status = submit(handle, request);
+            if (packet_acquire_status == PacketAcquireStatus.ConcurrencyMaxExceeded.value) {
+                throw new ConcurrencyExceededException();
+            } else if (packet_acquire_status == PacketAcquireStatus.Shutdown.value) {
+                throw new IllegalStateException("Client is closing");
+            } else {
+                assertTrue(packet_acquire_status == PacketAcquireStatus.Ok.value,
+                        "PacketAcquireStatus=%d is not implemented", packet_acquire_status);
+            }
+        } finally {
+            contextHandleReferences.decrementAndGet();
         }
     }
 
     public void close() {
-        if (contextHandle != 0L) {
+        if (contextHandle.getAcquire() != 0L) {
             synchronized (this) {
-                if (contextHandle != 0L) {
-                    // Deinit and signalize that this client is closed by setting the handles to 0.
-                    clientDeinit(contextHandle);
-                    this.contextHandle = 0L;
+                final var handle = contextHandle.getAcquire();
+                if (handle != 0L) {
+                    // Signalize that this client is closed by setting the handler to 0,
+                    // and spin wait until all references that might be using the old handle could
+                    // be released.
+                    contextHandle.setRelease(0L);
+                    while (contextHandleReferences.getAcquire() > 0L) {
+                        // Thread::onSpinWait method to give JVM a hint that the following code is
+                        // in a spin loop. This has no side-effect and only provides a hint to
+                        // optimize spin loops in a processor specific manner.
+                        Thread.onSpinWait();
+                    }
+
+                    // This function waits until all submited requests are completed, and no more
+                    // packets can be acquired after that.
+                    clientDeinit(handle);
```

Comment:
Heh, overall, I can't say I am too happy about this:

- abstractly, nesting synchronized and a spinloop (well, and spinloop by itself https://matklad.github.io/2020/01/02/spinlocks-considered-harmful.html :D) doesn't feel right. Ideally, I'd love to see something more principled here
- concretely, I think a flaw in this implementation is that, with many concurrent calls to close, only _one_ would actually wait for the stuff to get close. That is, it can be the case that a `.close()` call returns, but there's still some outstanding work. 

However, if I got this on the sync correctly, we are going to nix this code anyway soon, so I guess that's fine! Certainly better than an outright race we have at the moment!

Reply:
Yeah, my first thought was to encapsulate the context on a thread-safe class controlling all aspects more elegantly, but as you said, we are going to nuke this entire logic very soon.

About the `close` call, I don't think it can have side effects; the semantics are the same: only the first call to `close` will wait, no matter if concurrent or sequential calls.